src/sqswatcher/sqswatcher.py

```diff
@@ -223,51 +248,108 @@ def _retrieve_all_sqs_messages(queue, max_processed_messages):
+    return _resolve_events_hostname_collision(
+        event_pool=itertools.chain(parsed_events[EventType.ADD], parsed_events[EventType.REMOVE])
+    )
+
+
+def _parse_health_messages(sqs_config, messages, table, health_queue):
+    if messages:
+        instances_in_cluster = get_cluster_instance_info(
+            sqs_config.stack_name, sqs_config.region, sqs_config.proxy_config, include_master=False
+        )
+        parsed_events = _parse_messages_helper(
+            messages, sqs_config, queue_type=QueueType.health, table=table, instances_in_cluster=instances_in_cluster
+        )
+        return _resolve_events_hostname_collision(event_pool=parsed_events[EventType.HEALTH])
+    else:
+        return []
+
+
+def _parse_messages_helper(messages, sqs_config, queue_type, table=None, queue=None, instances_in_cluster=None):
+    parsed_events = {
+        EventType.ADD: [],
+        EventType.REMOVE: [],
+        EventType.HEALTH: [],
+    }
+    event_name_to_processing_function = {
+        EventType.HEALTH: lambda: _process_scheduled_maintenance_event(
+            message_attrs, message, table, instances_in_cluster
+        ),
+        EventType.ADD: lambda: _process_compute_ready_event(
+            sqs_config.region, sqs_config.proxy_config, message_attrs, message, table
+        ),
+        EventType.REMOVE: lambda: _process_instance_terminate_event(message_attrs, message, table, queue),
+    }
+    event_name_to_event_type = {
+        "parallelcluster:EC2_SCHEDULED_EVENT": EventType.HEALTH,
+        "parallelcluster:COMPUTE_READY": EventType.ADD,
+        "autoscaling:EC2_INSTANCE_TERMINATE": EventType.REMOVE,
+    }
+    for message in messages:
+        # CW event rule can only send messages to SQS with "" enclosing the message.
+        # However, when message is re-queued it is sent without ""
+        message_text = (
+            json.loads(json.loads(message.body)) if str(message.body).startswith('"') else json.loads(message.body)
+        )
+        message_attrs = json.loads(message_text.get("Message"))
+        event_name = message_attrs.get("Event")
+        if not event_name:
+            log.warning("Unable to read message. Deleting.")
+            message.delete()
+            continue
+        event_type = event_name_to_event_type.get(event_name, None)
+
+        if event_type in SUPPORTED_EVENTTYPE_FOR_QUEUETYPE[queue_type]:
+            event = event_name_to_processing_function.get(event_type)()
+            if event:
+                parsed_events[event_type].append(event)
+        else:
+            log.info("Unsupported event type %s for %s queue. Discarding message.", event_type, queue_type)
+            message.delete()
+
+    return parsed_events
+
+
+def _process_scheduled_maintenance_event(message_attrs, message, table, instances_in_cluster):
+    instance_id = message_attrs.get("EC2InstanceId")
+    if instance_id in instances_in_cluster:
+        hostname = _retrieve_hostname_from_ddb(instance_id, table)
```

Comment:
Probably we want to handle the case when hostname is None. Node could have been already removed. Shall we discard the message?